python/lsst/daf/butler/registry/interfaces/database.py

```diff
@@ -0,0 +1,927 @@
+    """
+
+    def __init__(self, db: Database):
+        self._db = db
+        self._foreignKeys = []
+        self._inspector = sqlalchemy.engine.reflection.Inspector(self._db._engine)
+        self._tableNames = frozenset(self._inspector.get_table_names(schema=self._db.namespace))
+
+    def addTable(self, name: str, spec: ddl.TableSpec) -> sqlalchemy.schema.Table:
+        """Add a new table to the schema, returning its sqlalchemy
+        representation.
+
+        The new table may not actually be created until the end of the
+        context created by `Database.declareStaticTables`, allowing tables
+        to be declared in any order even in the presence of foreign key
+        relationships.
+        """
+        if name in self._tableNames:
+            _checkExistingTableDefinition(name, spec, self._inspector.get_columns(name,
+                                                                                  schema=self._db.namespace))
+        table = self._db._convertTableSpec(name, spec, self._db._metadata)
+        for foreignKeySpec in spec.foreignKeys:
+            self._foreignKeys.append(
+                (table, self._db._convertForeignKeySpec(name, foreignKeySpec, self._db._metadata))
+            )
+        return table
+
+    def addTableTuple(self, specs: Tuple[ddl.TableSpec]) -> Tuple[sqlalchemy.schema.Table]:
+        """Add a named tuple of tables to the schema, returning their
+        SQLAlchemy representations in a named tuple of the same type.
+
+        The new tables may not actually be created until the end of the
+        context created by `Database.declareStaticTables`, allowing tables
+        to be declared in any order even in the presence of foreign key
+        relationships.
+        """
+        return specs._make(self.addTable(name, spec) for name, spec in zip(specs._fields, specs))
+
+
+class Database(ABC):
+    """An abstract interface that represents a particular database engine's
+    representation of a single schema/namespace/database.
+
+    Parameters
+    ----------
+    origin : `int`
+        An integer ID that should be used as the default for any datasets,
+        quanta, or other entities that use a (autoincrement, origin) compound
+        primary key.
+    namespace : `str`, optional
+        Name of the schema or namespace this instance is associated with.
+        This is passed as the ``schema`` argument when constructing a
+        `sqlalchemy.schema.MetaData` instance.  We use ``namespace`` instead to
+        avoid confusion between "schema means namespace" and "schema means
+        table definitions".
+    engine: `sqlalchemy.engine.Engine`.
+        SQLAlchemy engine object.  May be shared with other `Database`
+        instances.
+    connection : `sqlalchemy.engine.Connection`, optional
+        SQLAlchemy connection object.  May be shared with other `Database`
+        instances.  Constructed from ``engine`` if not provided.
```

Comment:
Still, if the same connection instance is shared between two databases then there is a potential that state of one Database instance (`_transactions` list) may not reflect actual state of the connection which was modified by other Database instance. Maybe instead of tracking transactions in `_transactions` list you could just check whether connection is already in transaction?

Reply:
Hmm, that's a good point.  I still think that this probably would be easy to avoid in the future `Registry` code that I expect to be the only place shared connections are used (other than unit tests), but this does feel like a problem in the interface if we have to rely on code outside of `Database` behaving a certain way in order for its behavior to be correct.  I'll see if I can guarantee that the `_transactions` list is shared between `Databases` if and only if the `_connection` is.

Reply:
I've pushed one more big commit, which adds some factory `classmethods` and a new `dataclass` that ensures the engine, connection and list of transactions is passed together.  It's probably still possible to misuse, but I think it's now a lot harder to *accidentally* misuse, and the checks for whether a transaction is active should now be safe.